### ISSUE TITLE
bug(types): type Inference for `decodeEventLog` not inferring the correct value

### ISSUE THREAD

**Issue by jxom** [wagmi-dev/viem#381]
### Discussed in https://github.com/wagmi-dev/viem/discussions/380

<div type='discussions-op-text'>

<sup>Originally posted by **mhmulford0** April 17, 2023</sup>
I have this bit of code spitting out all of the logs for a given ERC20 contract. Based on the docs for `decodeEventLog`, I should get a return type that looks like this:

```ts
/**
 *  {
 *    eventName: 'Transfer',
 *    args: {
 *      from: '0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266',
 *      to: '0x70997970c51812dc3a010c7d01b50e0d17dc79c8'
 *      value: 1n
 *    }
 *  }
 */
```

However, what I actually get is 
![Screenshot 2023-04-16 at 2 37 27 PM](https://user-images.githubusercontent.com/63870279/232337802-e12885fa-6516-43d6-8552-628cd5628e55.png)

This causes the required type assertion in the included code. After debugging a bit I could not get anything other than the empty object as the inferred type. Any help would be wonderful. Package version `0.3.2`

```ts
import { createPublicClient, decodeEventLog, http, parseAbiItem } from "viem";
import { mainnet } from "viem/chains";
import { erc20ABI } from "@wagmi/core";

type TransferLogArgs = {
  from: string;
  to: string;
  value: BigInt;
};

const transport = http(
  "https://eth-mainnet.g.alchemy.com/v2/my-correct-api-key"
);

const client = createPublicClient({
  chain: mainnet,
  transport,
});

async function main() {
  const BLOCK_INTERVAL = 250;
  let startBlock = 15390081;


  setInterval(async () => {
    console.log({ startBlock });
    const changes = await client.getLogs({
      address: "0xb24cd494faE4C180A89975F1328Eab2a7D5d8f11",
      event: parseAbiItem(
        "event Transfer(address indexed from, address indexed to, uint256 value)"
      ),
      fromBlock: BigInt(startBlock),
      toBlock: BigInt(startBlock + BLOCK_INTERVAL),
    });

    startBlock += BLOCK_INTERVAL;

    for (const change of changes) {
      const log = decodeEventLog({
        abi: erc20ABI,
        data: change.data,
        topics: change.topics,
        eventName: change.eventName,
      });

      if (typeof log.args === "undefined") {
        return new Error("incorrect data");
      }
  
      const info = log.args as TransferLogArgs;

      console.log(info);
    }
  }, 1500);
}

main().catch((err) => {
  console.log(err);
  process.exit(1);
});

```

**Comment by fubhy** [comment]
@tmm I got a bit stuck working on this. I think it's possible to strictly infer & enforce the `topics` as `readonly [Hex, ...Hex[]]` across the entire codebase.

Infering the right tuple shape should be as simple as this:

```typescript
type AbiEventParameter = AbiEvent['inputs'][number]

export type GetEventTopics<
  TAbi extends Abi | readonly unknown[],
  TEventName extends string,
  TAbiEvent extends AbiEvent & { type: 'event' } = TAbi extends Abi
    ? ExtractAbiEvent<TAbi, TEventName>
    : AbiEvent & { type: 'event' },
  TTopics extends readonly Hex[] = AbiEventInputToTopics<TAbiEvent['inputs']>,
> = readonly [Hex, ...TTopics]

type AbiEventInputToTopics<
  TAbiParameters extends readonly AbiEventParameter[],
> = TAbiParameters extends readonly [
  infer Head extends AbiEventParameter,
  ...infer Tail extends readonly AbiEventParameter[],
]
  ? Head extends { indexed: true }
    ? readonly [Hex, ...AbiEventInputToTopics<Tail>]
    : AbiEventInputToTopics<Tail>
  : []
```

However, that lead me down a deep rabbit trying to enforce those across the board as topics are rather losely typed atm.

Anyways... The root cause of the issue mentioned here is that `getLogs` already returns loosely typed `topics`. Passing that (needlessly) to `decodeEventLog` once again doesn't work then of course.

**Comment by fubhy** [comment]
Btw. related to this: Similarly, we could also more strictly infer the `data` portion of the logs here, making it required / optional as needed.

**Comment by jxom** [comment]
Yep. We the topics to match the length of indexed arguments + 1.

**Comment by fubhy** [comment]
> Yep. We need the inferred topics to match the length of indexed arguments + 1.

Yeah that's what this snippet does. But then I fell into the refactoring rabbit hole while trying to integrate this across the code base :-)

**Comment by jxom** [comment]
Fixed in #448
